### PR TITLE
Bare minimum support for gRPC-Web.

### PR DESCRIPTION
**Pull request by fengli79**
Bare minimum support for gRPC-Web.
https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md
1. Unary only.
2. No compression.

### CODE REVIEW

File: src/nginx/grpc_web_finish.cc [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+#include "src/nginx/grpc_finish.h"
+
+extern "C" {
+#include "src/http/ngx_http.h"
+}
+
+#define RETURN_IF_NULL(r, condition, ret, message)                  \
+  do {                                                              \
+    if (condition == nullptr) {                                     \
```
Comment (lizan) [inline src/nginx/grpc_web_finish.cc]: nit: (condition)

File: src/nginx/t/grpc_web_interop_status.t [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+################################################################################
+
+my $response = ApiManager::http($NginxPort,qq{
+POST /grpc.testing.TestService/UnaryCall HTTP/1.0
+Host: 127.0.0.1:${NginxPort}
+Content-Type: application/grpc-web
+x-api-key: api-key
+Content-Length: 15
+
+\x00\x00\x00\x00\x0a\x3a\x08\x08\x04\x12\x04\x74\x65\x73\x74});
```
Comment (lizan) [inline src/nginx/t/grpc_web_interop_status.t]: Can you add comments to note what those binaries are? (throughout)

File: src/nginx/grpc_web_finish.cc (hunk 1 above)
Comment (fengli79) [reply src/nginx/grpc_web_finish.cc]: Done

File: src/nginx/t/grpc_web_interop_status.t (hunk 2 above)
Comment (fengli79) [reply src/nginx/t/grpc_web_interop_status.t]: Done

File: src/nginx/t/grpc_web_interop_status.t [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+$t->run_daemon(\&ApiManager::grpc_interop_server, $t, "${GrpcBackendPort}");
+is($t->waitforsocket("127.0.0.1:${ServiceControlPort}"), 1, 'Service control socket ready.');
+is($t->waitforsocket("127.0.0.1:${GrpcBackendPort}"), 1, 'GRPC test server socket ready.');
+$t->run();
+is($t->waitforsocket("127.0.0.1:${NginxPort}"), 1, 'Nginx socket ready.');
+
+################################################################################
+
+my $response = ApiManager::http($NginxPort,qq{
+POST /grpc.testing.TestService/UnaryCall HTTP/1.0
```
Comment (qiwzhang) [inline src/nginx/t/grpc_web_interop_status.t]: It seems that all 4 "t" tests have same setup.  Only the requests and responses are different.   How about combine them into one "t" test? .

Comment (fengli79) [reply src/nginx/t/grpc_web_interop_status.t]: It follows the same pattern of grpc_interop_xxx tests. Prefer to keep them separate.

**Comment by ByteFlinger** [comment]
Is there any documentation on how to configure this?

I am running ESP in front of my grpc application and when I send a unary requests I receive the following answer back (Presumibly from endpoints)

```
{
 "code": 3,
 "message": "Unexpected token.\nAAAAAFIKQE5vcmRlbkFw\n^",
 "details": [
  {
   "@type": "type.googleapis.com/google.rpc.DebugInfo",
   "stackEntries": [],
   "detail": "internal"
  }
 ]
}
```

**Comment by qiwzhang** [comment]
The error indicates that request body is not a valid JSON.

**Comment by ByteFlinger** [comment]
I am just making a grpc-web request. From what I have seen the response body is a base64 value which should be correct.

Does grpc Google endpoints expect a JSON request?

**Comment by qiwzhang** [comment]
ESP grpc web support is not well documented.   It only supports following content type

  (content_type == "application/grpc-web" ||
   content_type == "application/grpc-web+proto")

The code is here:
https://github.com/cloudendpoints/esp/blob/74fefb91bee99282076c03ed6f122e7894a988cb/src/nginx/grpc.cc#L133

You probably uses grpc-web-text if you base64 encodes it.

**Comment by ByteFlinger** [comment]
Thank you for the heads up.

I was trying with the default client. I now see from the grpc-web documentation that basically this means that Google Endpoints does not support server streaming calls, only unary request-response grpc calls. Is that correct?

I hope support for that is added soon.

**Comment by qiwzhang** [comment]
Yeap,  ESP does't support grpc-web-text.

**Pull request merged**